File: dataset/output/GEMINI_gemini-1.5-pro-001/Task187_GEMINI_gemini-1.5-pro-001.cpp
#include <mutex>
#include <condition_variable>
#include <thread>

class H2O {
public:
    H2O() : hCount(0), oCount(0) {}

    void hydrogen(function<void()> releaseHydrogen) {
        unique_lock<mutex> lock(m);
        cv.wait(lock, [this] { return (hCount < 2 && oCount == 0) || (hCount == 2 && oCount == 1); });
        // releaseHydrogen() outputs "H". Do not change or remove this line.
        releaseHydrogen();
        hCount++;
        if (hCount == 2 && oCount == 1) {
            hCount = 0;
            oCount = 0;
        }
        cv.notify_all();
    }

    void oxygen(function<void()> releaseOxygen) {
        unique_lock<mutex> lock(m);
        cv.wait(lock, [this] { return hCount == 2 && oCount == 0; });
        // releaseOxygen() outputs "O". Do not change or remove this line.
        releaseOxygen();
        oCount++;
        cv.notify_all();
    }

private:
    mutex m;
    condition_variable cv;
    int hCount;
    int oCount;
};

File: dataset/output/GEMINI_gemini-1.5-pro-001/task187/Task187_GEMINI_gemini-1.5-pro-001.java
import java.util.concurrent.Semaphore;
import java.util.concurrent.CyclicBarrier;

class H2O {
    private Semaphore hSemaphore = new Semaphore(2);
    private Semaphore oSemaphore = new Semaphore(1);
    private CyclicBarrier barrier = new CyclicBarrier(3);

    public H2O() {
    }

    public void hydrogen(Runnable releaseHydrogen) throws InterruptedException {
        hSemaphore.acquire();
        try {
            barrier.await();
            // releaseHydrogen() outputs "H". Do not change or remove this line.
            releaseHydrogen.run();
        } finally {
            hSemaphore.release();
        }
    }

    public void oxygen(Runnable releaseOxygen) throws InterruptedException {
        oSemaphore.acquire();
        try {
            barrier.await();
            // releaseOxygen() outputs "O". Do not change or remove this line.
            releaseOxygen.run();
        } finally {
            oSemaphore.release();
        }
    }
}